README.md
[![CI Checks](https://github.com/Salable/errors/actions/workflows/ci.yml/badge.svg)](https://github.com/Salable/errors/actions/workflows/ci.yml)&nbsp;
[![License Checker](https://github.com/Salable/errors/actions/workflows/license-checker.yml/badge.svg)](https://github.com/Salable/errors/actions/workflows/license-checker.yml)&nbsp;
[![Release](https://github.com/Salable/errors/actions/workflows/release.yml/badge.svg)](https://github.com/Salable/errors/actions/workflows/release.yml)&nbsp;
<img src="https://img.shields.io/npm/v/@salable/errors">

# Salable Errors

The Salable Errors allows you to properly return HTTPException classes for specific error messages in response to HTTP requests to improve error handling in an application.

## Getting Started

<hr>

### Installation

> with npm

```bash
npm install @salable/errors
```

> with yarn

```bash
yarn add @salable/errors
```

### Using HTTPClientException - client errors

> With description

```ts
import {HTTPClientException} from '@salable/errors';
const errorPayload = {
  description:
    'Your changes were saved, but we could not connect your account due to a technical issue on our end. Please try connecting again.',
};
throw new HTTPClientException('NOT_FOUND', 'Unable to connect to account', errorPayload);
// Expected return
{
  statusCode: 404;
  message: 'Unable to connect to account';
  description: 'Your changes were saved, but we could not connect your account due to a technical issue on our end. Please try connecting again.';
}
```

> With NO description

```ts
import {HTTPClientException} from '@salable/errors';
throw new HTTPClientException('NOT_FOUND', 'Unable to connect to account');
// Expected return
{
  statusCode: 404;
  message: 'Unable to connect to account';
  description: undefined;
}
```

### Using HTTPServerException - Server Errors

> With description

```ts
import {HTTPServerException} from '@salable/errors';
const errorPayload = {
  description:
    'No changes were made, but we could not connect your account due to a technical issue on our end. Please try connecting again.',
};
throw new HTTPServerException('INTERNAL_SERVER_ERROR', 'Unable to connect to network', errorPayload);
// Expected return
{
  statusCode: 500;
  message: 'Unable to connect to account';
  description: 'No changes were made, but we could not connect your account due to a technical issue on our end. Please try connecting again.';
}
```

> With NO description

```ts
import {HTTPServerException} from '@salable/errors';
throw new HTTPServerException('INTERNAL_SERVER_ERROR', 'Unable to connect to network');
// Expected return
{
  statusCode: 500;
  message: 'Unable to connect to account';
  description: undefined;
}
```


CHANGELOG.md
## [1.0.3](https://github.com/Salable/errors/compare/v1.0.2...v1.0.3) (2023-04-14)


### Bug Fixes

* unexpected token 'export' error ([0a90760](https://github.com/Salable/errors/commit/0a9076035807f7cb92c7c8ae14217696dcb6c401))

## [1.0.2](https://github.com/Salable/errors/compare/v1.0.1...v1.0.2) (2023-04-13)


### Bug Fixes

* add description to error message ([d730524](https://github.com/Salable/errors/commit/d73052418405493678cd710dfb0ff5a816805b05))

## [1.0.1](https://github.com/Salable/errors/compare/v1.0.0...v1.0.1) (2023-04-13)


### Bug Fixes

* release package to registry ([dbf49d7](https://github.com/Salable/errors/commit/dbf49d70344c48951ab304a570539b3482622b31))

# 1.0.0 (2023-04-13)


### Features

* add http exceptions ([bcaa138](https://github.com/Salable/errors/commit/bcaa138bcd685e60b12e6627afe088242c3a6a10))


src/__tests__/http-exceptions.test.ts
import {HTTPClientException, HTTPServerException} from '../';

describe('HTTP Exception', () => {
  describe('Client Exception', () => {
    test('Should set properties correctly', () => {
      const errorPayload = {
        message: 'Unable to connect to account',
        description:
          'Your changes were saved, but we could not connect your account due to a technical issue on our end. Please try connecting again.',
      };
      const error = new HTTPClientException('NOT_FOUND', errorPayload.message, {description: errorPayload.description});
      expect(error.statusCode).toBe(404);
      expect(error.message).toBe(errorPayload.message);
      expect(error.description).toBe(errorPayload.description);
    });
    it('Should set default values correctly', () => {
      const message = 'Bad request';
      const error = new HTTPClientException('BAD_REQUEST', message);
      expect(error.statusCode).toBe(400);
      expect(error.message).toBe(message);
      expect(error.description).toBeUndefined();
    });
  });
  describe('Server Exception', () => {
    test('Should set properties correctly', () => {
      const errorPayload = {
        message: 'Unable to connect to network',
        description:
          'No changes were made, but we could not connect your account due to a technical issue on our end. Please try connecting again.',
      };
      const error = new HTTPServerException('INTERNAL_SERVER_ERROR', errorPayload.message, {
        description: errorPayload.description,
      });
      expect(error.statusCode).toBe(500);
      expect(error.message).toBe(errorPayload.message);
      expect(error.description).toBe(errorPayload.description);
    });
    it('Should set default values correctly', () => {
      const message = 'Internal server error';
      const error = new HTTPServerException('INTERNAL_SERVER_ERROR', message);
      expect(error.statusCode).toBe(500);
      expect(error.message).toBe(message);
      expect(error.description).toBeUndefined();
    });
  });
});


src/exceptions/client-http-exception.ts
import {IErrorBody} from './types';

enum ClientStatusCode {
  'BAD_REQUEST' = 400,
  'UNAUTHORIZED' = 401,
  'PAYMENT_REQUIRED' = 402,
  'FORBIDDEN' = 403,
  'NOT_FOUND' = 404,
  'METHOD_NOT_ALLOWED' = 405,
  'NOT_ACCEPTABLE' = 406,
  'REQUEST_TIMEOUT' = 408,
  'CONFLICT' = 409,
  'GONE' = 410,
  'LENGTH_REQUIRED' = 411,
  'PRECONDITION_FAILED' = 412,
  'PAYLOAD_TOO_LARD' = 413,
  'URI_TOO_LONG' = 414,
  'UNSUPPORTED_MEDIA_TYPE' = 415,
  'RANGE_NOT_SATISFIABLE' = 416,
  'EXPECTATION_FAILED' = 417,
  'I_AM_A_TEAPOT' = 418,
  'MISDIRECTED_REQUEST' = 421,
  'UNPROCESSABLE_CONTENT' = 422,
  'LOCKED' = 423,
  'FAILED_DEPENDENCY' = 424,
  'TOO_EARLY' = 425,
  'UPGRADE_REQUIRED' = 426,
  'PRECONDITION_REQUIRED' = 428,
  'TOO_MANY_REQUESTS' = 429,
  'REQUEST_HEADER_FIELDS_TOO_LARGE' = 431,
  'UNAVAILABLE_FOR_LEGAL_REASONS' = 451,
}
type StatusType = keyof typeof ClientStatusCode;
export class HTTPClientException extends Error {
  statusCode: number;
  description?: string;
  constructor(statusCode: StatusType, message: string, body?: IErrorBody) {
    super(message);
    this.statusCode = ClientStatusCode[statusCode];
    this.description = body?.description;
  }
}
